src/queue.cpp

```diff
@@ -970,7 +972,25 @@ cvk_command_kernel::build_batchable_inner(cvk_command_buffer& command_buffer) {
 
     // Setup descriptors
     if (!m_argument_values->setup_descriptor_sets()) {
-        return CL_OUT_OF_RESOURCES;
+        if (config.descriptor_sets_allocate_retry_sleep_us == UINT32_MAX) {
+            return CL_OUT_OF_RESOURCES;
+        }
+        // Retry every 'config.descriptor_set_allocate_retry_sleep_us' us until
+        // we have no more batch in flight, which would mean that all the
+        // descriptor should have been freed, thus the error is not about not
+        // having enough descriptors.
+        bool result_valid;
+        do {
+            m_queue->command_pool_unlock();
+            std::this_thread::sleep_for(std::chrono::microseconds(
+                config.descriptor_sets_allocate_retry_sleep_us));
+            m_queue->command_pool_lock();
+            result_valid = m_argument_values->setup_descriptor_sets();
+        } while (!result_valid && m_queue->have_batch_in_flight());
+        if (!result_valid) {
+            cvk_error_fn("could not setup descriptor sets");
+            return CL_OUT_OF_RESOURCES;
+        }
```

Comment:
I did wonder whether we could encapsulate the retry within `setup_descriptor_sets` but this would mean keeping the command pool locked for the whole duration of the retry, preventing other threads from enqueuing kernels from entry points that still have descriptor sets available in their pool. I think the performance advantage is worth the lack of encapsulation. We could probably restructure things to avoid that trade-off but that'd be way beyond the scope of this PR. In short: fine with your approach :).

Reply:
I have looked at it when making the PR. The issue is even worse, if the command pool is locked we cannot free descriptors set: `program.hpp: cvk_entry_point::free_descriptor_set` locks the command pool.

Reply:
Right, this is becoming a bit too tricky to reason about. We don't need to solve this here but I've created https://github.com/kpet/clvk/issues/658 so we don't forget about this issue.